python-api/gstswitch/server.py

```diff
@@ -22,6 +24,107 @@
 TOOLS_DIR = '/'.join(os.getcwd().split('/')[:-1]) + '/tools/'
 
 
+class ProcessMonitor(object):
+    """Runs a Command in a Background-Thread and monitors it's output
+
+    Can block until the Command prints a certain string and log the full
+    output into a file
+    """
+
+    def __init__(self, cmd, logtarget=sys.stderr):
+        self.log = logging.getLogger('server-output-monitor')
+
+        # After calling start(), _proc contains a subprocess.Popen instance
+        self._proc = None
+
+        # Command to run (Array passed to subprocess.Popen)
+        self._cmd = cmd
+
+        # Logfile to write to
+        self._logtarget = logtarget
+
+        # Internal Buffer to search for matched when wait_for_output is called
+        self._buffer = ""
+
+    def start(self):
+        """Starts up the process specified in the constructor in a thread.
+        Blocks until the process is started and has a pid assigned, but not
+        longer then 0.5s.
+        """
+
+        assert self._proc is None
+        self.log.debug("starting subprocess")
+
+        self._proc = subprocess.Popen(
+            self._cmd,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.STDOUT,
+            bufsize=0,
+            shell=False)
+
+        assert self._proc is not None
+        self.log.debug("subprocess successfully")
+
+    def terminate(self):
+        """Kills the process and waits for the thread to exit"""
+        self.log.debug("killing the subprocess")
+        if self._proc:
+            self._proc.terminate()
+
+    def poll(self):
+        """Calls poll() on the running process and returns its result or None
+        if no process is running"""
+        if self._proc:
+            return self._proc.poll()
+
+        return None
+
+    @property
+    def pid(self):
+        """Returns the pid of the running process or None
+        if no process is running"""
+        if self._proc:
+            return self._proc.pid
+
+        return None
+
+    def wait_for_output(self, match, timeout=5, count=1):
```

Comment:
But that doesn't work: calling `os.read` on a pipe that has no data will block that read forever (or until the subprocess actually outputs data). `os.read` does not take a timeout while `select.select` does.

The only way to avoid the select would be a reading thread and I worked on that for over a week now (see https://github.com/MaZderMind/gst-switch/blob/55910db80fc8471c4f0fb434c57cc426a496a7f2/python-api/gstswitch/server.py#L26) and it sometimes failes to read all of the subprocesses output for unknown reasons (~4% probability, did some statistics on it).

Honestly: I have spent over a week trying to find a reliable solution and I have one implemented right here. I'm not going to spend more time into fixing an unreliable one when I have one that succeeds 14000 runs with 100.0% probablility (14 tests ran 1000×), I'll opt for doing something productive like implementing the missing tests.